### ISSUE TITLE
[Bug]: NPU silently produces wrong predictions for Swin (window-attention) models — 0/500 agreement with the identical IR on CPU/GPU, compiles with no error and no fallback

### ISSUE THREAD

**Issue by cheng315ncu** [openvinotoolkit/openvino#37264]
### OpenVINO Version

2026.2.1

### Operating System

Other (Please specify in description)  — Ubuntu 24.04.2 LTS, kernel 6.14.0-37-generic

### Device used for inference

NPU

### Framework

PyTorch

### Model used

timm swin_tiny_patch4_window7_224.ms_in1k (public ImageNet-1k checkpoint, unmodified)

## Summary

On Intel NPU (Meteor Lake, device architecture `3720`), a stock Swin-Tiny classifier compiles
without any error or warning, reports pure NPU execution, and returns finite non-NaN logits —
but predicts a **different class than CPU/GPU on every single one of 500 test images**.

Nothing in the API signals that the result is wrong. `compile_model` succeeds,
`EXECUTION_DEVICES` reports `NPU` (no HETERO fallback), outputs are well-formed floats.

The same `.xml`/`.bin` pair, the same `ov.Core`, the same input tensors — only the device string
changes between the three runs below:

| model | weights | device | agreement with CPU | median logit range (max−min) | distinct predicted classes |
|---|---|---|---|---|---|
| ResNet-101 (`resnet101d`) — control | public timm | GPU | 500/500 (100 %) | — | — |
| ResNet-101 (`resnet101d`) — control | public timm | **NPU** | **500/500 (100 %)** | — | — |
| Swin-Tiny (`swin_tiny_patch4_window7_224`) | public timm | GPU | 498/500 (99.6 %) | 10.70 (CPU: 10.71) | 46 (CPU: 46) |
| Swin-Tiny (`swin_tiny_patch4_window7_224`) | public timm | **NPU** | **0/500 (0 %)** | **8.19** | **19** |

CPU vs GPU disagree on 2/500 images — ordinary fp16 cross-device rounding noise. CPU vs NPU
disagree on 500/500. GPU vs NPU also 0/500.

The NPU output is not merely different, it is **mode-collapsed**: over 1000 ImageNet classes, one
class is predicted for 282/500 images (56.4 %) and the top 4 classes cover 84.8 % of all
predictions, almost independent of the input image. The logit distribution is visibly compressed
(median range 8.19 vs 10.70 on CPU).

### Example — image 0 of the imagenette subset, public weights

```
CPU top-5: classes [  0, 758, 389, 764, 390]  logits [8.82, 4.20, 3.62, 2.43, 2.27]
NPU top-5: classes [475, 421, 506, 836, 928]  logits [4.68, 4.56, 3.98, 3.92, 3.75]
```

CPU has a clear winner (8.82 vs 4.20, Δ=4.62). NPU's top-5 is nearly flat (Δ=0.12 between 1st and
2nd) and shares **no class at all** with CPU's top-5.

### Also reproduces on a fine-tuned model

An internal 6-class Swin-Tiny classifier (same backbone, 6-class head) drops from **99.20 %
accuracy on CPU and GPU to 17.69 % on NPU** over its full 503-image validation set — and only
2 of the 6 classes are ever predicted, in roughly a 6:1 ratio regardless of true label. The same
ResNet-101 control on the same internal pipeline matches its CPU baseline exactly (501/503).

## Why this is NPU execution, not our export pipeline

1. **The convolutional control is clean.** ResNet-101, exported and converted through the
   *identical* script path, agrees 500/500 on NPU with public weights and matches CPU accuracy
   exactly on internal weights. The failure is architecture-selective.
2. **The cross-device baseline is three orders of magnitude smaller.** CPU vs GPU on this same
   Swin-Tiny IR: 2/500. NPU: 500/500.
3. **One file, one `read_model`, three device strings.** There is no NPU-specific export or
   conversion step — `core.compile_model(model, 'NPU')` is the only difference.
4. **Neither precision knob the NPU plugin exposes changes anything.**
   - `ov::hint::execution_mode = ACCURACY` → still wrong (sampled 17/101 correct on the internal model)
   - `ov::hint::inference_precision = f16` (explicit) → still wrong (19/101)
   - `ov::hint::inference_precision = f32` → rejected outright:
     `Failed to parse 'INFERENCE_PRECISION_HINT' option`. NPU 3720 has no fp32 compute path to
     fall back to, unlike the GPU plugin.
   - `ov::intel_npu::compiler_dynamic_quantization` and `qdq_optimization` →
     `NOT_FOUND for current configuration`. This is a plain FP16 graph with no
     QuantizeLinear/DequantizeLinear nodes, so no user-facing quantization switch is being
     silently defaulted on.
5. **Run-to-run jitter exists but is far too small to explain this.** Recompiling and rerunning
   the same image 3× gives logits varying by ~0.1
   (`[-0.457, 0.637, -0.642, 0.231, 1.049, 0.092]` vs `[-0.437, 0.635, -0.663, 0.231, 1.151, 0.081]`)
   — likely reduction-order effects. Each run's argmax is stable. An 82-point accuracy loss is
   not a rounding artifact.

## `NPU_TILES=1` trades this for a hard compile failure

Forcing single-tile compilation (disabling the default 2-cluster tiling) does **not** fix the
accuracy — it makes the model fail to compile entirely, with a different error:

```
Got Diagnostic at loc(.../layers.0/blocks/blocks.1/attn/softmax/Softmax) :
  failed to legalize operation 'builtin.unrealized_conversion_cast'
Failed Pass ConvertViewOpsToDeclarations
```

This is the **same error signature at the same op type** (window-attention `Softmax`) that
`NPU_TILES=1` produces on Swin-Large — which does not compile under *any* tiling configuration
(filed separately as a companion report, linked in a comment below). Across both model sizes and
both tiling configurations, every
failure we found is anchored inside a window-attention block: silent value corruption at the
default multi-tile schedule, hard MLIR legalization failure at `Softmax` under single-tile.

We could not find a configuration in which a Swin model both compiles and produces correct output
on this NPU.

## What we are asking for

1. **Is silently-wrong output for window-attention models a known limitation** of NPU 3720 / this
   compiler version? The public NPU-device documentation lists static shapes as a limitation, but
   documents no accuracy caveat for attention-heavy architectures — unlike, for example, the GPU
   fp16 accuracy note.
2. **If known:** please document it explicitly, and if possible add a runtime warning or an
   `ov::hint::execution_mode=ACCURACY`-gated check. The core hazard here is that
   **compiling without error is not a signal of correctness** — a user deploying Swin to NPU today
   gets plausible-looking, completely wrong results with nothing to alert them.
3. **If it is a genuine bug:** any pointer to which part of the window-attention lowering is the
   likely culprit (shifted-window partition/reverse, relative-position-bias add, or the `Softmax`
   itself, given where the `NPU_TILES=1` crash lands) would help us narrow it further.

### Step-by-step reproduction

Fully self-contained — public timm weights and imagenette2-320, no private data.
Requires an Intel NPU (Meteor Lake or newer) with the NPU driver installed, and membership in the
`render` group for `/dev/accel0`.

**1. Environment**

```bash
python -m venv ov-repro && source ov-repro/bin/activate
pip install "openvino==2026.2.1" timm torch numpy pillow
```

**2. Data** (imagenette2-320, ~340 MB, no auth)

```bash
wget https://s3.amazonaws.com/fast-ai-imageclas/imagenette2-320.tgz
tar -xzf imagenette2-320.tgz
```

**3. Export the model to ONNX, then to FP16 IR**

```python
# export.py
import timm, torch
m = timm.create_model('swin_tiny_patch4_window7_224.ms_in1k', pretrained=True).eval()
with torch.no_grad():
    torch.onnx.export(m, torch.zeros(1, 3, 224, 224), 'swin_tiny.onnx',
                      input_names=['input'], output_names=['logits'],
                      opset_version=17, dynamo=False)
```

```bash
python export.py
ovc swin_tiny.onnx --output_model swin_tiny_fp16 --compress_to_fp16 True
```

**4. Compare the same IR across CPU / GPU / NPU**

```python
# compare.py
import numpy as np, openvino as ov, timm
from timm.data import resolve_data_config, create_transform
from PIL import Image
from pathlib import Path

# 500 images: 50 per class, round-robin over the 10 imagenette classes
root = Path('imagenette2-320/val')
classes = sorted(p for p in root.iterdir() if p.is_dir())
per = [sorted(c.glob('*.JPEG'))[:50] for c in classes]
files = [per[c][i] for i in range(50) for c in range(len(classes))]

cfg = resolve_data_config({}, model=timm.create_model(
    'swin_tiny_patch4_window7_224.ms_in1k', pretrained=False))
cfg['input_size'] = (3, 224, 224)
tfm = create_transform(**cfg)
x = np.stack([tfm(Image.open(f).convert('RGB')).numpy() for f in files]).astype(np.float32)

core = ov.Core()
model = core.read_model('swin_tiny_fp16.xml')
name = model.input(0).get_any_name()

preds = {}
for dev in ('CPU', 'GPU', 'NPU'):
    comp = core.compile_model(model, dev)
    print(dev, 'EXECUTION_DEVICES =', comp.get_property('EXECUTION_DEVICES'))
    preds[dev] = np.array([
        np.argmax(list(comp({name: x[i:i+1]}).values())[0][0]) for i in range(len(x))])

for dev in ('GPU', 'NPU'):
    agree = int((preds['CPU'] == preds[dev]).sum())
    print(f'CPU vs {dev}: {agree}/{len(x)} agree  '
          f'({len(set(preds[dev].tolist()))} distinct classes predicted)')
```

**Expected:** NPU agrees with CPU/GPU to within ordinary fp16 cross-device noise — as GPU does
here (498/500).

**Actual:**

```
CPU EXECUTION_DEVICES = ['CPU']
GPU EXECUTION_DEVICES = ['GPU.0']
NPU EXECUTION_DEVICES = ['NPU']
CPU vs GPU: 498/500 agree  (46 distinct classes predicted)
CPU vs NPU: 0/500 agree    (19 distinct classes predicted)
```

**5. Control — swap in ResNet-101 to confirm the pipeline is sound**

Replace the model id with `resnet101d.ra2_in1k` in both scripts and rerun: NPU agrees 500/500.

### Relevant log output

```
# Environment
Host          : Intel Core Ultra 7 155H (Meteor Lake), Intel(R) AI Boost NPU, architecture 3720
OS            : Ubuntu 24.04.2 LTS, kernel 6.14.0-37-generic
NPU driver    : intel-driver-compiler-npu / intel-level-zero-npu / intel-fw-npu
                1.10.0.20241107-11729849322
                NPU_DRIVER_VERSION property = 1731009178
NPU compiler  : NPU_COMPILER_TYPE=DRIVER (install default), NPU_COMPILER_VERSION=393217
OpenVINO      : 2026.2.1 (PyPI), Python 3.12
Model         : timm swin_tiny_patch4_window7_224.ms_in1k
                torch.onnx.export (opset 17, static [1,3,224,224]) -> ovc --compress_to_fp16
Data          : imagenette2-320 val, 500 images stratified 50 x 10

# Result — no error or warning is emitted at any point
CPU EXECUTION_DEVICES = ['CPU']
GPU EXECUTION_DEVICES = ['GPU.0']
NPU EXECUTION_DEVICES = ['NPU']
CPU vs GPU: 498/500 agree  (46 distinct classes predicted)
CPU vs NPU: 0/500 agree    (19 distinct classes predicted)

# Per-image example (image 0)
CPU top-5: classes [  0, 758, 389, 764, 390]  logits [8.82, 4.20, 3.62, 2.43, 2.27]
NPU top-5: classes [475, 421, 506, 836, 928]  logits [4.68, 4.56, 3.98, 3.92, 3.75]

# NPU_TILES=1 — compile failure instead of wrong output
Got Diagnostic at loc(.../layers.0/blocks/blocks.1/attn/softmax/Softmax) :
  failed to legalize operation 'builtin.unrealized_conversion_cast'
Failed Pass ConvertViewOpsToDeclarations

# Rejected precision hint
ov::hint::inference_precision = f32
  -> Failed to parse 'INFERENCE_PRECISION_HINT' option

# Not applicable (plain FP16 graph, no Q/DQ nodes)
ov::intel_npu::compiler_dynamic_quantization -> NOT_FOUND for current configuration
ov::intel_npu::qdq_optimization             -> NOT_FOUND for current configuration
```

### Issue submission checklist

- [x] I'm reporting an issue. It's not a question.
- [x] I checked the problem with the documentation, FAQ, open issues, Stack Overflow, etc., and have not found a solution.
- [x] There is reproducer code and related data files such as images, videos, models, etc.

---

**Comment by cheng315ncu** [comment]
The companion report referenced above is now filed as #37265 — Swin-Large fails to compile on NPU with an MLIR legalization failure at a residual `Add` in the same window-attention block.

The two are separable but likely share a root cause: this issue is silent value corruption under the default multi-tile schedule, #37265 is a hard compile failure, and forcing `NPU_TILES=1` on *either* model lands on the same `unrealized_conversion_cast` legalization failure at the window-attention `Softmax`. That may be worth one investigation rather than two.

**Comment by cheng315ncu** [comment]
Resolved on my side — this is fixed by the NPU driver, with OpenVINO unchanged. Posting the numbers in case they help anyone who lands here from a search.

**What changed:** only the NPU driver stack, upgraded `1.10.0.20241107-11729849322` → `1.24.0.20251003-18218973328` (`intel-driver-compiler-npu`, `intel-fw-npu`, `intel-level-zero-npu`). OpenVINO stayed at `2026.2.1-21919` and the device is the same 3720. Since `NPU_COMPILER_TYPE=DRIVER`, graph compilation happens in the driver's compiler rather than in OpenVINO, which is consistent with the same OpenVINO build behaving differently across the upgrade.

| | driver 1.10.0 (as filed) | driver 1.24.0 |
|---|---|---|
| public `swin_tiny_patch4_window7_224.ms_in1k`, CPU vs NPU | **0/500** agree | **500/500** agree |
| same, CPU vs GPU (control) | 498/500 | 498/500 — unchanged |
| same, distinct predicted classes on NPU | mode-collapsed to a small subset | 46, identical to CPU/GPU |
| same, median logit range (max−min) on NPU | visibly compressed | 10.698 (CPU 10.705, GPU 10.695) |
| internal 6-class Swin-Tiny, full 503-image val | 99.20 % → **17.69 %** | **99.20 %**, 1.0000 agreement with the PyTorch reference — which CPU also scores, so CPU and NPU agree on every one of the 503 images |
| `NPU_DRIVER_VERSION` | 1731009178 | 1759485217 |
| `NPU_COMPILER_VERSION` | 393217 | 458777 |

The re-measurement used byte-identical IR, cached inputs and comparison scripts. The CPU-vs-GPU control reproducing exactly at 498/500 is what rules out an accidental setup difference between the two sessions.

One corroborating detail suggesting the old driver was genuinely at fault rather than the original numbers being mismeasured: shortly before the upgrade, still on 1.10.0, the kernel logged

```
intel_vpu 0000:00:0b.0: [drm] *ERROR* ivpu_mmu_dump_event(): MMU EVTQ: 0x10 (Translation fault)
    SSID: 2 SID: 3, e[2] 00000000, e[3] 00000208, in addr: 0x1deaddea0, fetch addr: 0x0
intel_vpu 0000:00:0b.0: [drm] *ERROR* ivpu_pm_trigger_recovery(): Recovery triggered by TDR
```

The `deaddea` pattern in that fault address looks like a memory poison value, i.e. freed or uninitialised memory being dereferenced.

Two things that might still be worth acting on, if maintainers agree:

1. The original ask in this report was not only "fix the wrong output" but "make silent wrongness detectable". A miscompilation that returns finite, plausible logits with `EXECUTION_DEVICES` reporting clean NPU execution is invisible to a caller. Whatever changed between these driver versions, the *class* of failure — compiles fine, runs fine, wrong answer — remains undetectable at the API level.
2. If the fix is identifiable, a note in the NPU device docs giving the first driver version with correct window-attention output would let people on older drivers know they need to upgrade rather than assume their model is at fault.

Happy to close this from my side. The companion #37265 (Swin-Large fails to compile) **still reproduces on 1.24.0** — I've commented there separately.
